### ISSUE TITLE
Replace one column with another in bash

### ISSUE THREAD

**Issue by yaaminiv** [RobertsLab/resources#488]
I am performing a flanking analysis to understand where DML and DMR overlap with 1000 bp regions upstream or downstream from mRNA coding regions in [this Jupyter notebook](
https://github.com/fish546-2018/yaamini-virginica/blob/master/notebooks/2018-11-01-DML-and-DMR-Analysis.ipynb). I used `flankBed` to add 1000 bp flanks upstream and downstream of each [mRNA coding region](
http://eagle.fish.washington.edu/Cvirg_tracks/C_virginica-3.0_Gnomon_mRNA.gff3). My goal now is to isolate the flanks themselves, then use `intersectBed` to identify overlaps between the flanks and DML/DMR.

This is the approach I'm considering for isolating those 1000 bp flanks:

1. Create a dataframe with the "upstream flank": This dataframe will have the chromosome number, mRNA indication, start bp (from the extended flank), stop bp (from the original mRNA GFF), and all other information from the original document.
2. Create a similar dataframe with the "downstream flank": This dataframe will have the start bp from the original mRNA GFF and the stop bp from the extended flank.
3. Merge the "upstream flank" and "downstream flank" data frames together, then use `intersectBed`

I know I can isolate my intended start and stop bp columns from their respective dataframes. What bash commands can I use to essentially cut and paste columns into a new dataframe? I figure I'll need to use `awk`, but I don't know how........

**Comment by yaaminiv** [comment]
Here's what my dataframes look like:

mRNA coding region GFF:

<img width="774" alt="screen shot 2018-11-15 at 11 12 25 am" src="https://user-images.githubusercontent.com/22335838/48575661-62e34e80-e8c7-11e8-9a40-c2f6efee7f7c.png">

mRNA coding regions with flanks:

<img width="769" alt="screen shot 2018-11-15 at 11 11 56 am" src="https://user-images.githubusercontent.com/22335838/48575653-61b22180-e8c7-11e8-9ddd-24cb2072682f.png">

**Comment by kubu4** [comment]
All of your links actually link to this GtiHub issue instead of their intended targets.

An example of a before and your desired after of your dataframe would be very helpful.

**Comment by yaaminiv** [comment]
@kubu4 That issue should be fixed now! I opened the issue before I finished adding in all of the links. I also added images of the dataframes I'm working with and my desired dataframe.

**Comment by kubu4** [comment]
Can you run please run ```head -n 1``` ? I think it will be easier to read. Also, can you please provide column names (doesn't have to be in file, just listed out in order that they exist in your dataframes)?

**Comment by yaaminiv** [comment]
@kubu4 Sure!

mRNA off:

<img width="773" alt="screen shot 2018-11-15 at 11 16 28 am" src="https://user-images.githubusercontent.com/22335838/48575976-2bc16d00-e8c8-11e8-8f6a-8b8f241c93cf.png">

mRNA gff with 1000 bp flanks on both ends:

<img width="776" alt="screen shot 2018-11-15 at 11 17 24 am" src="https://user-images.githubusercontent.com/22335838/48575977-2c5a0380-e8c8-11e8-8fa2-535de5ccc2c7.png">

Column names (applies to both dataframes):

1. Chromosome
2. "Gnomon"
3. "mRNA"
4. Start bp
5. Stop bp
6. "."
7. Strand
8. "."
9. Gene name and annotation information

I would like to retain all of the columns in my final dataframe(s)

**Comment by yaaminiv** [comment]
To clarify, these are the column names I want:

Upstream flanks:

1. Chromosome
2. "Gnomon"
3. "mRNA"
4. Start bp (column 4 from mRNA gff with 1000 bp flanks)
5. Stop bp (column 4 from mRNA gff)
6. "."
7. Strand
8. "."
9. Gene name and annotation information

Downstream flanks:

1. Chromosome
2. "Gnomon"
3. "mRNA"
4. Start bp (column 5 from mRNA gff)
5. Stop bp (column 5 from mRNA gff with 1000 bp flanks)
6. "."
7. Strand
8. "."
9. Gene name and annotation information

**Comment by kubu4** [comment]
What are the file source(s) for columns 1-3, 6-9 in your final output files?

**Comment by kubu4** [comment]
Something like this (haven't tested):

UPSTREAM FLANKS
```
paste \
<(cut -f-3 {mRNAList}) \
<(cut -f4 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed) \
<(cut -f4 {mRNAList}) \
<(cut -f6-9 {mRNAList})
```

DOWNSTREAM FLANKS
```
paste \
<(cut -f-3 {mRNAList}) \
<(cut -f5 {mRNAList}) \
<(cut -f5 2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed) \
<(cut -f6-9 {mRNAList})
```

And, to explain all of this...

```paste``` joins together input data. Default delimiter is tab. So, it will paste together all of the outputs from each of the following ```cut``` commands.

The ```cut``` commands cut based on tabs (default delimiter) and fields. Field selections are specified by ```-f``` and can use a single column (e.g. ```cut -f5```) or a range of columns (e.g. ```cut -f-3```; this cuts from the beginning of the document to field 3).

The ```<()``` notation is bash thing (process substitution) for storing the output of a command without needing an intermediate variable and when using a piper (```|```) isn't feasible.

**Comment by yaaminiv** [comment]
Thanks for the explanation! I tried the upstream flank command and got the following error:

<img width="777" alt="screen shot 2018-11-15 at 1 08 46 pm" src="https://user-images.githubusercontent.com/22335838/48581798-a9d94000-e8d7-11e8-9684-b60194501f23.png">

And a similar error with the downstream flank command:

<img width="773" alt="untitled" src="https://user-images.githubusercontent.com/22335838/48581847-d1c8a380-e8d7-11e8-9ba1-0b2268af9aa0.png">

Is there an issue with the column range...?

**Comment by sr320** [comment]
you need a 
`!` or `%%bash` ?

**Comment by kubu4** [comment]
Exactly what @sr320 said.

**Comment by yaaminiv** [comment]
Whoops... :expressionless:

I still do have an issue. The command is unable to find my variable {mRNAlist} in the third and fourth lines of the code:

<img width="778" alt="untitled" src="https://user-images.githubusercontent.com/22335838/48582347-21f43580-e8d9-11e8-9c36-a08cb8db50ab.png">

I obviously could just use the relative path to the file, but I already have the variable set. Why can't I use it?

**Comment by yaaminiv** [comment]
@kubu4 I actually didn't have a problem with it not recognizing `2018-11-14-Flanking-Analysis/2018-11-14-mRNA-1000bp-Flanks.bed`! I didn't close/re-open the notebook, and my `cd` is exactly what I want it to be so all of my relative paths work.

I reran the command using both the relative path to the GFF and the variable path and didn't get an error when I used the relative path. I'll continue to just use that relative path for now (v frustrating).

And yes! I did realize I need to redirect output. Just was curious about what was going on before redirecting output :)

**Comment by kubu4** [comment]
Interesting. I interpreted the second line of the errors above as coming from the flanking bed file, since it wasn't preceded by the ```cut: {mRNAList}```.

**Comment by yaaminiv** [comment]
No, I think that it was connected to the `cut: {mRNAList}` since there were two `cut: {mRNAList}` followed by two `No such file or directory`

In any case, the code works. Thanks!

**Comment by kubu4** [comment]
Looking at your notebook, I see that your variables all work and you use the ```!``` to run them. Have you tried using that with the ```paste``` command above? Just curious if that has any effect. No biggie if you don't want to bother checking it out.

**Comment by yaaminiv** [comment]
I tried using `!`, but encountered a syntax error (even with relative paths):

<img width="770" alt="screen shot 2018-11-15 at 3 38 34 pm" src="https://user-images.githubusercontent.com/22335838/48588583-8c62a100-e8ec-11e8-84fa-1d66a15c4633.png">

**Comment by kubu4** [comment]
Thanks for testing!

**Comment by yaaminiv** [comment]
PLOT TWIST:

I went through the commands and was looking at the output, and it was pretty funky. Looking at `head` for the original mRNA gff and the flanking document I realized that `bedtools flank` ALREADY ISOLATES THE FLANK FOR YOU (smh :expressionless:)

<img width="783" alt="untitled" src="https://user-images.githubusercontent.com/22335838/48589280-b79abf80-e8ef-11e8-978c-c97bc596cf05.png">

This wasn't very clear to me when I read the manual. Now it is.

Turns out I didn't need this code today, but at least I learned something! #themoreyouknow :rainbow:
